src/training.py

```diff
@@ -87,17 +104,122 @@ def training_step(self, batch, batch_idx):
         )
         return loss
 
+    def compute_metrics(self, embeddings: torch.Tensor, labels: torch.Tensor, clusters: np.ndarray) -> Dict[str, float]:
+        """
+        Compute various internal and external clustering metrics.
+
+        Parameters
+        ----------
+        embeddings : torch.Tensor
+            The embeddings on which clustering is performed.
+        labels : torch.Tensor
+            The actual labels for the embeddings.
+        clusters : np.ndarray
+            The clusters assignment from clustering algorithm.
+
+        Returns
+        -------
+        Dict[str, float]
+            A dictionary containing the computed metrics.
+        """
+
+        metrics = {}
+        averages = ['macro', 'micro', 'weighted']
+
+        metrics_functions = {
+            'internal': {
+                'silhouette_score': silhouette_score,
+                'davies_bouldin_score': davies_bouldin_score,
+                'calinski_harabasz_score': calinski_harabasz_score,
+            },
+            'external': {
+                'accuracy_score': accuracy_score,
+                **{
+                    f'{average}_{metric}': partial(metric_func, average=average, zero_division=0)
+                    for metric, metric_func in {
+                        'precision_score': precision_score,
+                        'recall_score': recall_score,
+                        'f1_score': f1_score,
+                    }.items()
+                    for average in averages
+                }
+            }
+        }
+
+        embeddings = embeddings.cpu().detach().numpy()
+        labels = labels.cpu().numpy()
+        for metric_name, metric_function in metrics_functions['internal'].items():
+            metrics['int_'+metric_name] = metric_function(embeddings, labels)
+
+        for metric_name, metric_function in metrics_functions['external'].items():
+            metrics['ext_'+metric_name] = metric_function(labels, clusters)
+
+        return metrics
+
+    def clustering(self, embeddings: np.ndarray, labels: np.ndarray) -> np.ndarray:
+        """
+        Perform clustering on the given embeddings using KMeans and map the labels according to maximum set intersection.
+
+        Parameters
+        ----------
+        embeddings : np.ndarray
+            The embeddings on which clustering is to be performed.
+        labels : np.ndarray
+            The actual labels for the given embeddings.
+
+        Returns
+        -------
+        np.ndarray
+            The labels predicted by KMeans mapped according to maximum intersection with actual labels.
+        """
+
+        kmeans = KMeans(n_clusters=40, random_state=0, n_init=10).fit(embeddings)
```

Comment:
make please `n_clusters = np.unique(labels).size` or create a separate param with explicit defining of clusters number